### PR TITLE
Clarify use of regex in access control configs

### PR DESCRIPTION
**Pull request by goakley**
Provide more details on what exactly is and isn't treated as regex.

### CODE REVIEW

**Comment by ptgott** [comment]
Hi @goakley, sorry for the long wait on this one. If you have time to address the conflicts and double-check the accuracy, I think this could still be useful.

**Comment by zmb3** [comment]
While accurate, this behavior looks to be a bug and not intentional. All selectors should be treated as regular expressions.

What's happening in this case is `GlobToRegex` is changing the input of `test|staging` to `test\|staging`, and then we surround that with `^` and `$` to get the final expression of `$test\|staging^`, which doesn't match.

File: docs/pages/access-controls/reference.mdx [hunk 1]
```diff
@@ -260,6 +262,10 @@ spec:
 
 ### Extended Node labels syntax
 
+Any string starting with `^` and ending with `$` will be interpreted as a regex and not as a literal string.
```
Comment (zmb3) [inline docs/pages/access-controls/reference.mdx:265]:
```suggestion
Any string starting with `^` or ending with `$` will be interpreted as a regex and not as a literal string.
```

It's technically not an and, either one of these conditions alone is enough for us to treat it as a regex.

File: docs/pages/access-controls/reference.mdx [hunk 2]
```diff
@@ -260,6 +262,10 @@ spec:
...
+For example, specifying a node label `environment: 'test|staging'` will only match the exact environment string "test|staging",
+while `environment: '^test|staging$'` will match either the string "test" or "staging".
```
Comment (zmb3) [inline docs/pages/access-controls/reference.mdx:267]: We should also mention that for strings that do not start with `^` or end with `$`, an `*` can be used as a glob-style matcher that matches any string.

**Review by zmb3 — COMMENTED** [review]
This behavior looks like it might have been an unintentional bug. I believe the original intent was that all selectors would be treated as regular expressions, because we pass all of them to a function called `SliceMatchesRegex` without trying to determine if it is a regex or not:

https://github.com/gravitational/teleport/blob/769951fe4b383db8ac8efcddb430e1a41dc970f7/lib/services/role.go#L1028-L1043

What's happening here is that case is that we're trying to simultaneously support both:
- glob-style matches like `cluster-*` (where the * matches any string)
- proper regular expressions

If we don't see a leading `^` or a trailing `$`, we assume it's a glob-style expression, and try to replace the `*` with proper regex syntax.

https://github.com/gravitational/teleport/blob/769951fe4b383db8ac8efcddb430e1a41dc970f7/lib/utils/replace.go#L83-L86

This changes the input regex from `test|staging` to `test\|staging` (via a call to `QuoteMeta`), making the regex look for a literal `|` and breaking the match.

https://github.com/gravitational/teleport/blob/769951fe4b383db8ac8efcddb430e1a41dc970f7/lib/utils/replace.go#L34

Given that this has been the behavior for several years, we should keep it as is. I would only request that we also mention the glob-style matching in this PR as well (see comments for specifics).

**Comment by ptgott** [comment]
@goakley Just checking whether you are planning to respond to @zmb3's comments. We can also take this one on internally

**Comment by ptgott** [comment]
Closing as stale. We'll pick this up again in the future via #19488.

**Pull request closed without merging**